server/pulp/server/controllers/repository.py

```diff
@@ -1007,3 +1035,423 @@ def _now_timestamp():
+
+    for content_unit in content_units:
+        # All files in the unit; every request for a unit has a reference to this dict.
+        unit_files = {}
+        unit_working_dir = os.path.join(working_dir, content_unit.id)
+        for file_path in content_unit.list_files():
+            qs = model.LazyCatalogEntry.objects.filter(
+                unit_id=content_unit.id,
+                unit_type_id=content_unit.type_id,
+                path=file_path
+            )
+            catalog_entry = qs.order_by('revision').first()
+            if catalog_entry is None:
+                continue
+            signed_url = _get_streamer_url(catalog_entry, signing_key)
+
+            temporary_destination = os.path.join(
+                unit_working_dir,
+                os.path.basename(catalog_entry.path)
+            )
+            mkdir(unit_working_dir)
+            unit_files[temporary_destination] = {
+                CATALOG_ENTRY: catalog_entry,
+                PATH_DOWNLOADED: None,
+            }
+
+            request = DownloadRequest(signed_url, temporary_destination)
+            # For memory reasons, only hold onto the id and type_id so we can reload the unit
+            # once it's successfully downloaded.
+            request.data = {
+                TYPE_ID: content_unit.type_id,
+                UNIT_ID: content_unit.id,
+                UNIT_FILES: unit_files,
+            }
+            requests.append(request)
+
+    return requests
+
+
+def _get_streamer_url(catalog_entry, signing_key):
+    """
+    Build a URL that can be used to retrieve the file in the catalog entry from
+    the lazy streamer.
+
+    :param catalog_entry: The catalog entry to get the URL for.
+    :type  catalog_entry: pulp.server.db.model.LazyCatalogEntry
+    :param signing_key: The server private RSA key to sign the url with.
+    :type  signing_key: M2Crypto.RSA.RSA
+
+    :return: The signed streamer URL which corresponds to the catalog entry.
+    :rtype:  str
+    """
+    try:
+        https_retrieval = parse_bool(pulp_conf.get('lazy', 'https_retrieval'))
+    except Unparsable:
+        raise PulpCodedTaskException(error_codes.PLP1014, section='lazy', key='https_retrieval',
+                                     reason=_('The value is not boolean'))
+    retrieval_scheme = 'https' if https_retrieval else 'http'
+    host = pulp_conf.get('lazy', 'redirect_host')
+    port = pulp_conf.get('lazy', 'redirect_port')
+    path_prefix = pulp_conf.get('lazy', 'redirect_path')
+    netloc = (host + ':' + port) if port else host
+    path = os.path.join(path_prefix, catalog_entry.path.lstrip('/'))
+    unsigned_url = urlunsplit((retrieval_scheme, netloc, path, None, None))
+    # Sign the URL for a year to avoid the URL expiring before the task completes
+    return str(URL(unsigned_url).sign(signing_key, expiration=31536000))
+
+
+class LazyUnitDownloadStep(DownloadEventListener):
```

Comment:
I filed https://pulp.plan.io/issues/1508 in response to this problem. The whole class needs to be refactored, but since we want a beta in a few minutes...